File: src/web/plugins/pcat.js
// DOS/V (standard PC, KVM) front-end plugin for the Mirai98 web manager.
//
// Registers the "pcat" machine: a plain PC/AT that QEMU runs under KVM for a
// DOS/V Windows 9x guest.  It shares the "dosv" Storage shelf with box86
// (the same floppies and CDs; a plain raw hard disk of its own).  Everything
// the console needs -- picture, sound over the VNC stream, media swap over
// QMP -- is the core's QEMU path, so there is no console hook here: only the
// create/edit form, the read-only hardware card, and one hard-disk format.

const PCAT_MEMS = ['64M', '128M', '256M', '512M'];
const PCAT_VGAS = [['std', 'std (Bochs VBE) -- SoftGPU'],
                   ['cirrus', 'Cirrus -- stock Win98 driver']];
// boot values match FM TOWNS' shared "boot" field (hd/fd/cd); net is a core
// field whose values are empty (isolated) or nat
const PCAT_BOOTS = [['hd', 'Hard disk'], ['fd', 'Floppy A'],
                    ['cd', 'CD-ROM']];
const PCAT_NETS = [['', 'Isolated (no network)'],
                   ['nat', 'NAT (QEMU user, RTL8139)']];

// The fixed parts, named so the create wizard shows what the machine really
// has where the stock "Sound/BIOS" selects would otherwise sit.
const PCAT_BIOS = 'QEMU SeaBIOS (pc-i440fx, ACPI off)';
const PCAT_SOUND = 'Sound Blaster 16 (ISA); Win98 has the driver';
if (typeof JA === 'object') {
  JA[PCAT_BIOS] = 'QEMU SeaBIOS (pc-i440fx, ACPI 無効)';
  JA[PCAT_SOUND] = 'Sound Blaster 16 (ISA); Win98 標準ドライバ';
}

function pcatEditForm(i, h) {
  const note = (t) => ' <span class="note">' + t + '</span>';
  const opts = (list, cur) => list.map(([v, l]) =>
    '<option value="' + v + '"' + ((cur || '') === v ? ' selected' : '') +
    '>' + h.esc(l) + '</option>').join('');
  const mems = PCAT_MEMS.includes(i.memory) ? PCAT_MEMS
    : PCAT_MEMS.concat([i.memory]);
  const memOpts = mems.map(m =>
    '<option' + (i.memory === m ? ' selected' : '') + '>' + m +
    '</option>').join('');
  return '<form onsubmit="return saveVm(this,\'' + i.name + '\')">' +
    '<div class="row"><label>Hard disk</label>' +
      h.diskSelect('hdd1', 'hdd', i.hdd1, null, 'dosv') +
      note('IDE primary master; a plain .img (raw) or .qcow2') + '</div>' +
    '<div class="row"><label>Hard disk 2</label>' +
      h.diskSelect('hdd2', 'hdd', i.hdd2, null, 'dosv') + '</div>' +
    '<div class="row"><label>CD-ROM</label>' +
      h.diskSelect('cd', 'cdrom', i.cd, null, 'dosv') +
      note('IDE secondary master; can be swapped while running') + '</div>' +
    '<div class="row"><label>Floppy A</label>' +
      h.diskSelect('fdd1', 'fdd', i.fdd1, null, 'dosv') + '</div>' +
    '<div class="row"><label>Floppy B</label>' +
      h.diskSelect('fdd2', 'fdd', i.fdd2, null, 'dosv') + '</div>' +
    '<div class="row"><label>Video</label>' +
      '<select name="vga">' + opts(PCAT_VGAS, i.vga || 'std') + '</select>' +
      note('std (Bochs VBE) for the SoftGPU driver; Cirrus for the stock ' +
           'Win98 display driver') + '</div>' +
    '<div class="row"><label>Boot from</label>' +
      '<select name="boot">' + opts(PCAT_BOOTS, i.boot || 'hd') +
      '</select></div>' +
    '<div class="row"><label>Network</label>' +
      '<select name="net">' + opts(PCAT_NETS, i.net || '') + '</select>' +
      note('isolated by default; NAT gives the guest outbound through QEMU ' +
           'with no host bridge') + '</div>' +
    '<div class="row"><label>Machine type</label>' +
      '<select name="machine">' + h.machineList().map(m =>
        '<option value="' + m + '"' +
        ((i.machine || 'pcat') === m ? ' selected' : '') + '>' +
        h.esc(h.machineLabel ? h.machineLabel(m) : m) +
        '</option>').join('') + '</select></div>' +
    '<div class="row"><label>Memory</label>' +
      '<select name="memory">' + memOpts + '</select></div>' +
    '<div class="row"><label>BIOS</label>' + note(PCAT_BIOS) +
      '<input type="hidden" name="bios" value="real"></div>' +
    '<div class="row"><label>Sound</label>' + note(PCAT_SOUND) +
      '<input type="hidden" name="sound" value="none"></div>' +
    '<div class="row"><label>Execution</label>' +
      '<label class="check"><input type="checkbox" name="kvm"' +
      (i.accel === 'kvm' ? ' checked' : '') +
      '> run on the host CPU (KVM), fall back to translation</label>' +
      note('KVM runs the guest near native; if the host cannot, it falls ' +
           'back to translation on its own') + '</div>' +
    '<div class="row"><label>Snapshot</label>' +
      '<label class="check"><input type="checkbox" name="snapshot"' +
      (i.snapshot ? ' checked' : '') + '> discard changes</label></div>' +
    '<div class="row"><label>Extra args</label>' +
      '<input type="text" name="extra" value="' + h.esc(i.extra || '') +
      '"></div>' +
    '<div class="row"><label></label><button class="primary"' +
      (i.running ? ' disabled title="stop it first"' : '') + '>Save</button>' +
      '<button type="button" onclick="removeVm(\'' + i.name + '\')"' +
      (i.running ? ' disabled' : '') + '>Delete</button></div></form>';
}

// The read-only "Hardware configuration" table.  The machine's own board,
// CPU, video and memory come from pcat.py's machine_shown (i.hardware), so
// the accelerator row can show what is really running (KVM or the TCG it
// fell back to) rather than only what was asked for.
function pcatHardware(i, h) {
  const spec = i.hardware || {};
  const boot = (PCAT_BOOTS.find(([v]) => v === (i.boot || 'hd')) ||
                ['', i.boot || 'hd'])[1];
  const net = (PCAT_NETS.find(([v]) => v === (i.net || '')) ||
               ['', i.net || 'isolated'])[1];
  const rows = [
    ['&#9881; Machine', h.esc(spec.machine || 'PC/AT (i440FX)')],
    ['&#9636; CPU', h.esc(spec.cpu || 'Pentium III')],
    ['&#9889; Acceleration',
     h.esc(spec.accel || (i.accel === 'kvm' ? 'KVM' : 'TCG'))],
    ['&#9635; Video', h.esc(spec.video || i.vga || 'std')],
    ['&#9737; Memory', h.esc(spec.memory || i.memory || '')],
    ['&#9834; Sound', h.esc(spec.sound || 'Sound Blaster 16')],
    ['&#9707; Hard disk', i.hdd1 ? h.esc(i.hdd1) : '(empty)']];
  if (i.hdd2) rows.push(['&#9707; Hard disk 2', h.esc(i.hdd2)]);
  rows.push(['&#9707; CD-ROM', i.cd ? h.esc(i.cd) +
    ' <span class="note">(read-only)</span>' : '(empty)']);
  rows.push(['&#9707; Floppy A', i.fdd1 ? h.esc(i.fdd1) : '(empty)']);
  if (i.fdd2) rows.push(['&#9707; Floppy B', h.esc(i.fdd2)]);
  rows.push(['&#9654; Boot from', h.esc(boot)]);
  rows.push(['&#8646; Network', h.esc(net)]);
  if (i.ports) rows.push(['&#9635; Display', 'VNC :' + (i.ports[0] - 5900) +
    ', websocket ' + i.ports[1]]);
  if (i.snapshot)
    rows.push(['&#8635; Snapshot', 'changes discarded on shutdown']);
  if (i.extra) rows.push(['&#9656; Extra args', h.esc(i.extra)]);
  return { rows, bios: PCAT_BIOS, sound: PCAT_SOUND };
}

// ---- create wizard --------------------------------------------------------
// The stock panes are PC-98 shaped: Disks lists the flat pc98 shelf, Network
// offers a bridge, and Host's shared folder / serial / GP-IB mean nothing
// here.  So Disks is drawn on the dosv shelf, Host/Memory/Network are dropped
// (Memory comes from the defaults above), and the machine's own choices go on
// an Options pane.
function pcatWizardDisks(h) {
  return '<div class="row"><label>Hard disk</label>' +
      h.diskSelect('hdd1', 'hdd', '', null, 'dosv') +
      h.note('a blank raw disk from Storage, or leave it empty and add one ' +
             'in Edit') + '</div>' +
    '<div class="row"><label>Floppy A</label>' +
      h.diskSelect('fdd1', 'fdd', '', null, 'dosv') + '</div>' +
    '<div class="row"><label>Floppy B</label>' +
      h.diskSelect('fdd2', 'fdd', '', null, 'dosv') + '</div>' +
    '<div class="row"><label>CD-ROM</label>' +
      h.diskSelect('cd', 'cdrom', '', null, 'dosv') + '</div>' +
    '<div class="note">Images live in Storage, on the shared dosv shelf. ' +
    'The CD and floppies can be swapped live from the detail page later; ' +
    'the hard disk needs the machine stopped.</div>';
}

function pcatWizardOptions(h) {
  const opts = (list, cur) => list.map(([v, l]) =>
    '<option value="' + v + '"' + ((cur || '') === v ? ' selected' : '') +
    '>' + h.esc(l) + '</option>').join('');
  return '<div class="row"><label>Video</label>' +
      '<select name="vga">' + opts(PCAT_VGAS, 'std') + '</select>' +
      h.note('std for the SoftGPU driver; Cirrus for the stock one') +
      '</div>' +
    '<div class="row"><label>Boot from</label>' +
      '<select name="boot">' + opts(PCAT_BOOTS, 'hd') + '</select></div>' +
    '<div class="row"><label>Network</label>' +
      '<select name="net">' + opts(PCAT_NETS, '') + '</select>' +
      h.note('isolated by default; NAT gives outbound through QEMU') +
      '</div>' +
    '<div class="row"><label>Execution</label>' +
      '<label class="check"><input type="checkbox" name="kvm" checked> ' +
      'run on the host CPU (KVM), fall back to translation</label></div>' +
    '<div class="row"><label>Snapshot</label>' +
      '<label class="check"><input type="checkbox" name="snapshot"> ' +
      'discard changes</label></div>';
}

function pcatWizardConfirm(v, h) {
  const pick = (list, cur, dflt) =>
    (list.find(([x]) => x === (cur || dflt)) || ['', cur || dflt])[1];
  const rows = [
    ['Name', h.esc(v.name || '(unnamed)')],
    ['Machine type', 'DOS/V PC (KVM, std VGA)'],
    ['Video', pick(PCAT_VGAS, v.vga, 'std')],
    ['Acceleration', v.kvm ? 'KVM' : 'TCG'],
    ['Boot from', pick(PCAT_BOOTS, v.boot, 'hd')],
    ['Network', pick(PCAT_NETS, v.net, '') || 'Isolated']];
  for (const [k, label] of [['hdd1', 'Hard disk'], ['fdd1', 'Floppy A'],
                            ['fdd2', 'Floppy B'], ['cd', 'CD-ROM']])
    if (v[k]) rows.push([label, h.esc(v[k])]);
  rows.push(['Snapshot', v.snapshot ? 'yes' : 'no']);
  return rows;
}

window.registerMachinePlugin({
  machines: ['pcat'],
  // its disks live on the shared dosv shelf, beside box86's
  platform: 'dosv',
  defaults: {
    pcat: { memory: '256M', vga: 'std', boot: 'hd', net: '',
            accel: 'kvm', sound: 'none', bios: 'real',
            lockSound: true, lockBios: true }
  },
  badge: { pcat: 'KVM' },
  labels: { pcat: 'DOS/V PC (KVM, std VGA)' },
  editForm: { pcat: pcatEditForm },
  hardware: { pcat: pcatHardware },
  // Storage "Create": a blank raw hard disk for this machine.  The dosv
  // floppy layouts and the CD come from box86's own registration on the
  // same shelf, so they are not repeated here.
  diskFormats: {
    hdd: [{ value: 'pcat-raw', label: 'PC/AT hard disk (blank, raw)',
            note: 'a sparse raw image of the size you give (in MB); name it ' +
                  '.img so it is read as a raw disk, then let Win98 ' +
                  'FDISK/FORMAT partition it, as on a real machine' }]
  },
  wizard: { pcat: { panes: { Disks: pcatWizardDisks, Host: null,
                             Memory: null, Network: null,
                             Options: pcatWizardOptions },
                    confirm: pcatWizardConfirm } }
});

File: src/plugins/pcat.py
# DOS/V (standard PC, KVM) machine plugin for the Mirai98 web manager.
#
# Registers the "pcat" machine: a plain PC/AT (QEMU i440FX + PIIX) run under
# KVM for a DOS/V Windows 9x guest, the same shape as the PC-98 and FM TOWNS
# machines -- a machine_argv builder, and the core does everything else
# (start/stop, the VNC console, QMP media swap, and the audio the VNC server
# carries).  It shares the "dosv" Storage shelf with box86; only the hard
# disk differs (a plain raw image here, not box86's 86Box VHD).
#
# Nothing in box86.py or pc98web.py is touched: a QEMU machine needs no
# engine of its own, so add_machine + machine_argv is the whole of it, plus
# the read-only hardware card (machine_shown) and one blank-disk format.

# The BIOS this boots is the QEMU fork's own pc-bios (bios-256k.bin,
# vgabios-*.bin); the fork does not put a standard-PC BIOS on its default
# search path, so the builder names it.  Overridable from mirai98.json
# ("pc_bios") without a code change.
PCAT_PC_BIOS = "/opt/mirai98-src/qemu-pc98towns/pc-bios"

PCAT_MACHINE_LABEL = "PC/AT (QEMU i440FX + PIIX)"
# SoftGPU's Win98 driver is an SSE3 build, so the vCPU advertises SSE3; the
# host has it under KVM and it is emulated under TCG.
PCAT_CPU_LABEL = "Pentium III (pentium3, +SSE/SSE2/SSE3)"
PCAT_SOUND_LABEL = "Sound Blaster 16 (ISA)"
PCAT_VGA_LABELS = {"std": "Bochs VBE (std)", "cirrus": "Cirrus CLGD5446"}

_BOOT_LETTER = {"hdd": "c", "fdd": "a", "cd": "d"}


def register(api):
    api.add_machine("pcat", platform="dosv")
    api.machine_argv("pcat", lambda inst: pcat_argv(api, inst))
    api.machine_shown("pcat", lambda inst: pcat_hardware(api, inst))
    # the machine-specific choices, so the record keeps them; unknown
    # values are turned back rather than silently kept
    api.add_field("vga", lambda v: None if v in ("", "std", "cirrus")
                  else "unknown video card")
    api.add_field("boot", lambda v: None if v in ("", "hdd", "fdd", "cd")
                  else "unknown boot device")
    api.add_field("net", lambda v: None if v in ("", "off", "nat")
                  else "unknown network mode")
    # a plain raw hard disk of this machine's own; box86's VHD is for
    # 86Box, and the one dosv shelf holds both, told apart by format
    api.disk_builder("dosv", "hdd", "pcat-raw", pcat_new_hard_disk)


def pcat_argv(api, inst):
    os = api.os
    cfg = api.CONFIG
    # ports_of hands back a fourth port (box86's own audio websocket);
    # a QEMU machine carries its audio on the VNC stream, so it is dropped
    vnc, ws, qmp_port, _audio_ws = api.ports_of(inst)
    display = vnc - 5900
    # KVM when asked, translation when not -- and, spelled kvm:tcg, KVM
    # that falls back to translation instead of refusing to start when the
    # host cannot give it
    accel = "kvm:tcg" if inst.get("accel", "kvm") == "kvm" else "tcg"
    host = "127.0.0.1" if api.LOOPBACK else "0.0.0.0"
    pcbios = cfg.get("pc_bios") or PCAT_PC_BIOS
    vga = inst.get("vga") or "std"
    boot = _BOOT_LETTER.get(inst.get("boot") or "hdd", "c")
    argv = [
        cfg["qemu"],
        # the standard-PC BIOS and vgabios first, then the fork's own data
        "-L", api.win_short(pcbios),
        "-L", api.win_short(cfg["datadir"]),
        # ACPI off for now: this Win98 SE goes through APM, and the ACPI
        # path is still being brought up on the isolated bench
        "-M", "pc-i440fx-9.2,acpi=off,accel=%s" % accel,
        "-cpu", "pentium3,+sse2,+sse3,+ssse3",
        "-m", inst.get("memory") or "256M",
        # the calendar clock follows the host's local time, so the guest
        # shows the wall clock rather than UTC
        "-rtc", "base=localtime",
        # No -k: this guest was installed with a 101 (US) keyboard, so
        # QEMU's default en-us VNC keymap matches it.  PC-98's own -k ja is
        # for its JIS keyboard and would put the wrong characters here.
        "-vga", vga,
        "-display", "none",
        # SB16 plays into a null backend; the VNC server captures that mix
        # (audiodev=snd on -vnc) and streams it to the browser, the same
        # path PC-98 and Towns sound takes
        "-audiodev", "none,id=snd",
        "-device", "sb16,audiodev=snd",
        "-vnc", "%s:%d,websocket=%d,audiodev=snd" % (host, display, ws),
        "-qmp", "tcp:127.0.0.1:%d,server=on,wait=off" % qmp_port,
        "-boot", "order=%s" % boot,
    ]
    if inst.get("snapshot"):
        argv.append("-snapshot")
    # IDE: primary master/slave are the hard disks (index 0/1), the CD is
    # the secondary master (index 2).  drive_backing reads the file's own
    # format, so a .qcow2 comes up as qcow2 and a .img/.raw as raw.
    for index, key in enumerate([k for k in ("hdd1", "hdd2") if inst.get(k)]):
        argv += ["-drive", "if=ide,index=%d,%s"
                 % (index, api.drive_backing(api.disk_path(inst, key)))]
    # the CD drive exists even with an empty tray, so a disc can be put in
    # from the Media row (over QMP) without a restart
    cd = "if=ide,index=2,media=cdrom,readonly=on"
    if inst.get("cd"):
        cd += "," + api.drive_backing(api.disk_path(inst, "cd"))
    argv += ["-drive", cd]
    # both floppy drives always exist, for the same reason as the CD
    for index, key in enumerate(("fdd1", "fdd2")):
        drive = "if=floppy,index=%d" % index
        if inst.get(key):
            drive += "," + api.drive_backing(api.disk_path(inst, key))
        argv += ["-drive", drive]
    if inst.get("net") == "nat":
        # QEMU's own NAT behind an RTL8139, which Win98 has a driver for;
        # nothing is asked of the host, so this needs no bridge
        argv += ["-netdev", "user,id=lan",
                 "-device", "rtl8139,netdev=lan"]
    if inst.get("extra"):
        argv += inst["extra"].split()
    return argv


def pcat_hardware(api, inst):
    """The read-only hardware card's facts, put on the record as its own
    `hardware` field.  A QEMU machine keeps no config file -- the record
    and the argv are the configuration -- so the board, CPU, video and
    memory come straight from the instance's fields and the fixed machine.

    The one thing worth asking the running machine is which accelerator it
    actually got: `accel=kvm:tcg` falls back to translation when KVM is
    unavailable, and the card should say what is really running rather than
    what was asked for.
    """
    vga = inst.get("vga") or "std"
    out = {
        "machine": PCAT_MACHINE_LABEL,
        "cpu": PCAT_CPU_LABEL,
        "video": PCAT_VGA_LABELS.get(vga, vga),
        "memory": inst.get("memory") or "256M",
        "sound": PCAT_SOUND_LABEL,
    }
    requested = "KVM" if inst.get("accel", "kvm") == "kvm" else "TCG"
    accel = None
    if api.is_running(inst):
        reply = api.qmp(inst, "query-kvm")
        if reply and isinstance(reply.get("return"), dict):
            enabled = reply["return"].get("enabled")
            if enabled is True:
                accel = "KVM (host CPU)"
            elif enabled is False:
                # asked for KVM, running translated -- the fallback fired
                accel = "TCG (translated%s)" % (
                    ", KVM unavailable" if requested == "KVM" else "")
        if accel is None:
            accel = requested + " (running)"
    else:
        accel = requested
    out["accel"] = accel
    return out


def _parse_size(text):
    """Bytes from a size like "2G" / "512M"; 2 GB when it cannot be read."""
    text = str(text).strip().upper()
    mult = 1
    if text.endswith("G"):
        mult, text = 1 << 30, text[:-1]
    elif text.endswith("M"):
        mult, text = 1 << 20, text[:-1]
    elif text.endswith("K"):
        mult, text = 1 << 10, text[:-1]
    try:
        return int(float(text) * mult)
    except ValueError:
        return 2 << 30


def pcat_new_hard_disk(dest, data):
    """A blank hard disk for the PC/AT machine: a sparse raw image, sized
    from the create form (2 GB by default).  Win98's own FDISK/FORMAT
    partitions and formats it, as on a real machine.  Kept raw (not qcow2)
    so no qemu-img is needed to make one; drive_backing reads it as raw.
    """
    size = _parse_size(data.get("size") or "2G")
    with open(dest, "wb") as f:
        f.truncate(size)
